### PR TITLE
snap: overhaul validation error messages

### PR DESCRIPTION
**Pull request by bboozzoo**
Overhaul error messages from snap validation in the hope of making them more useful by adding context.

### CODE REVIEW

File: snap/validate.go [hunk 1]
```diff
@@ -270,12 +271,12 @@ func validateSocketAddr(socket *SocketInfo, fieldName string, address string) er
...
 func validateSocketAddrPath(socket *SocketInfo, fieldName string, path string) error {
 	if clean := filepath.Clean(path); clean != path {
-		return fmt.Errorf("socket %q has invalid %q: %q should be written as %q", socket.Name, fieldName, path, clean)
+		return fmt.Errorf("invalid %q: %q should be written as %q", fieldName, path, clean)
 	}
 
 	if !(strings.HasPrefix(path, "$SNAP_DATA/") || strings.HasPrefix(path, "$SNAP_COMMON/")) {
 		return fmt.Errorf(
-			"socket %q has invalid %q: only $SNAP_DATA and $SNAP_COMMON prefixes are allowed", socket.Name, fieldName)
+			"invalid %q: only $SNAP_DATA and $SNAP_COMMON prefixes are allowed", fieldName)
```
Comment (chipaca) [inline snap/validate.go]: I've got a problem with this one: `only FOO and BAR prefixes are allowed` does not mean quite the same as `must have a prefix of FOO or BAR`, which is the case here. Granted this one was wrong before :-)

File: snap/validate.go [hunk 2]
```diff
@@ -310,13 +311,13 @@ func validateSocketAddrNetHost(socket *SocketInfo, fieldName string, address str
 		}
 	}
 
-	return fmt.Errorf("socket %q has invalid %q address %q", socket.Name, fieldName, address)
+	return fmt.Errorf("invalid %q address %q", fieldName, address)
```
Comment (chipaca) [inline snap/validate.go]: maybe add `(only addresses referring to localhost are valid)`? or is that explained elsewhere?

File: snap/validate.go [hunk 3]
```diff
@@ -304,19 +305,20 @@ func validateSocketAddrNet(socket *SocketInfo, fieldName string, address string)
...
-		if address == validAddress {
+	validAddresses := []string{"127.0.0.1", "[::1]", "[::]"}
+	for _, valid := range validAddresses {
+		if address == valid {
 			return nil
 		}
 	}
 
-	return fmt.Errorf("socket %q has invalid %q address %q", socket.Name, fieldName, address)
+	return fmt.Errorf("invalid %q address %q, must be one of: %s", fieldName, address, strings.Join(validAddresses, ", "))
```
Comment (stolowski) [inline snap/validate.go]: Nice.

File: snap/validate.go [hunk 4]
```diff
@@ -562,20 +564,18 @@ func validateAppOrderCycles(apps map[string]*AppInfo) error {
...
+		return errors.New("must be a service to define before/after ordering")
 	}
 
 	for _, dep := range dependencies {
 		// dependency is not defined
 		other, ok := app.Snap.Apps[dep]
 		if !ok {
-			return fmt.Errorf("application %q refers to missing application %q in before/after",
-				app.Name, dep)
+			return fmt.Errorf("refers to missing application %q in before/after", dep)
```
Comment (stolowski) [inline snap/validate.go]: +1 for changing the original message, but the new version feels like it's missing a subject. How about: `reference to a missing application ....`? Perhaps @degville  can advise again.

File: snap/validate.go [hunk 5]
```diff
@@ -562,20 +564,18 @@ func validateAppOrderCycles(apps map[string]*AppInfo) error {
...
 		}
 
 		if !other.IsService() {
-			return fmt.Errorf("application %q refers to non-service application %q in before/after",
-				app.Name, dep)
+			return fmt.Errorf("refers to non-service application %q in before/after", dep)
```
Comment (stolowski) [inline snap/validate.go]: Same.

File: snap/validate.go (hunk 4 above)
Comment (degville) [reply snap/validate.go]: +1 on @stolowski's suggestion, or how about `before/after references a missing application`?

File: snap/validate.go (hunk 5 above)
Comment (degville) [reply snap/validate.go]: `before/after references a non-service application, %q` ... ?

**Pull request merged**